### DISCUSSION TITLE
How to distinguish ferrite and bainite with KAM

### DISCUSSION THREAD

**Discussion by Mikewangkang** [AzdiarGazder/mtexTools#4] [Q&A] [answered]
hello @AzdiarGazder 
I need help with a Matlab® code developed that uses the MTEX toolbox to first define grain boundaries based on selected grain tolerance angles and then calculate the KAM for each pixel. If the misorientation of a pixel exceeds 3°, the program classifies the grains as granular bainite; otherwise, it is classified as polygonal ferrite. The classification algorithm then color-codes the relative sorted grains it determines and calculates the measured granular bainite and polygonal fractions. This is my purpose, I have proceeded to calculate the kam value of each pixel but the follow-up process I don't know how to do.
I would be very grateful if anyone could help!
Wang Kang

**Comment by AzdiarGazder** [comment, marked as answer]
Hi @Mikewangkang,

There are various methods used to classify and segment ferrite constituents. I am providing 2 examples as follows:
1. One method, which I developed, can be found [here](https://www.sciencedirect.com/science/article/pii/S0304399114001284).
2. Another method, developed by Sachin Shrestha, can be found [here](https://www.sciencedirect.com/science/article/pii/S0304399113002957).

You can apply my method by going through the various steps listed in the paper.
Alternatively, I have modified Sachin's method to work within MTEX.
The demonstration script can be found [here](https://github.com/AzdiarGazder/mtexTools/tree/main/ferriteQuantifier).

I'd like to highlight a few important points here:
(a) Based on the information you've provided, you seem to suggest that 1 criterion (like KAM) is enough to segment between constituents. However, modern segmentation methods, like the two I've shown above, use multiple criteria to separate between the constituents. Using only 1 criterion is simply not enough to accurately (and consistently) segment between constituents.

(b) A criterion like KAM is a pixel-based criterion. When segmenting constituents, it is advisable to use grain-based criteria like GAM and GOS. This enables cleaner and easier segmentation.

(c) Depending on the thermo-mechanical history of each individual microstructure (and consequently, the map data), the thresholds should be adjusted/changed to accurately (and consistently) segment between constituents. Using 1 threshold value across all microstructures and maps is again simply not correct.


Lastly, if you'd like me to help you with your script, please upload it so that I can have a look.

Hope this helps.

Warm regards,
Azdi

**Comment by Mikewangkang** [comment]
Dear Professor Azdi,
Thank you very much for your reply, I will study your method well.
Best wishes,
wang  kang

**Comment by Mikewangkang** [comment]
Dear Professor Azdi,
The following is my script. I can calculate the kam diagram, but then I don’t know how to write the script to realize that the kam value greater than 3° is defined as grain shell and others are classified as polygonal ferrite. Can you help me?

```matlab
%% Import Script for EBSD Data
%
% This script was automatically created by the import wizard. You should
% run the whoole script or parts of it in order to import your data. There
% is no problem in making any changes to this script.

%% Specify Crystal and Specimen Symmetries

% crystal symmetry
CS = {... 
  'notIndexed',...
  crystalSymmetry('m-3m', [2.9 2.9 2.9], 'mineral', 'Iron bcc (old)', 'color', [0.53 0.81 0.98]),...
  crystalSymmetry('m-3m', [3.7 3.7 3.7], 'mineral', 'Iron fcc', 'color', [0.56 0.74 0.56])};

% plotting convention
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','intoPlane');

%% Specify File Names

% path to files
pname = 'D:\桌面\晶体塑性模拟\B钢\B钢EBSD\B750';

% which files to be imported
fname = [pname '\B750-1.crc'];

%% Import the Data

% create an EBSD variable containing the data
ebsd = EBSD.load(fname,CS,'interface','crc',...
  'convertEuler2SpatialReferenceFrame');

% The second parameter that is involved in grain reconstruction is the threshold misorientation angle indicating a grain boundary.
grains = calcGrains(ebsd,'angle',5*degree)

[grains,ebsd.grainId] = calcGrains(ebsd('indexed'));

% remove one-three pixel grains
ebsd(grains(grains.grainSize <= 3)) = [];
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'));

grains = smooth(grains,5);

plot(ebsd('indexed'),ebsd('indexed').orientations)
hold on
plot(grains.boundary,'lineWidth',1.5)
hold off

ebsd = ebsd.gridify;

kam = ebsd.KAM / degree;

% lets plot it
plot(ebsd,kam,'micronbar','off')
caxis([0,15])
mtexColorbar
mtexColorMap LaboTeX
hold on
plot(grains.boundary,'lineWidth',1.5)
hold off

% chose a denoising filter
F = halfQuadraticFilter;
F.alpha = 0.5;

% denoise the orientation map
ebsdS = smooth(ebsd,F,'fill',grains);

% plot the first order KAM
plot(ebsdS,ebsdS.KAM('threshold',2.5*degree) ./ degree,'micronbar','off')
caxis([0,5])
mtexColorbar
mtexColorMap LaboTeX
hold on
plot(grains.boundary,'lineWidth',1.5)
hold off
```

Best wishes,
wang kang
